ext/js/dictionary/dictionary-data-util.js

```diff
@@ -146,23 +146,20 @@ export class DictionaryDataUtil {
 
         /** @type {Map<string, import('dictionary-data-util').GroupedPronunciationInternal[]>} */
         const groupedPronunciationsMap = new Map();
-        for (const {headwordIndex, dictionary, pitches} of pronunciations) {
+        for (const {headwordIndex, dictionary, pronunciations} of termPronunciations) {
             const {term, reading} = headwords[headwordIndex];
             let dictionaryGroupedPronunciationList = groupedPronunciationsMap.get(dictionary);
             if (typeof dictionaryGroupedPronunciationList === 'undefined') {
                 dictionaryGroupedPronunciationList = [];
                 groupedPronunciationsMap.set(dictionary, dictionaryGroupedPronunciationList);
             }
-            for (const {position, nasalPositions, devoicePositions, tags} of pitches) {
-                let groupedPronunciation = this._findExistingGroupedPronunciation(reading, position, nasalPositions, devoicePositions, tags, dictionaryGroupedPronunciationList);
+            for (const pronunciation of pronunciations) {
+                let groupedPronunciation = this._findExistingGroupedPronunciation(reading, pronunciation, dictionaryGroupedPronunciationList);
                 if (groupedPronunciation === null) {
                     groupedPronunciation = {
+                        ...pronunciation,
```

Comment:
Can this be made to not be a rest expansion? I.e. just `pronunciation`?

Reply:
Not sure what you mean, like this?
```
export type GroupedPronunciation = {
    pronunciation: GroupedPitchAccentInternal | GroupedPhoneticTranscriptionInternal;
    exclusiveTerms: string[];
    exclusiveReadings: string[];
};
```

Reply:
Correct. Expanding it with `...` seems like it should be unnecessary and would just create a more complex object to construct.

Reply:
Sorry, I can't get the types to work when it's like that.

Reply:
```patch
diff --git a/ext/js/data/sandbox/anki-note-data-creator.js b/ext/js/data/sandbox/anki-note-data-creator.js
index e4e0b64d..f27cee86 100644
--- a/ext/js/data/sandbox/anki-note-data-creator.js
+++ b/ext/js/data/sandbox/anki-note-data-creator.js
@@ -196,9 +196,11 @@ export class AnkiNoteDataCreator {
             for (const {dictionary, pronunciations} of DictionaryDataUtil.getGroupedPronunciations(dictionaryEntry)) {
                 /** @type {import('anki-templates').Pitch[]} */
                 const pitches = [];
-                for (const pronunciation of pronunciations) {
+                for (const groupedPronunciation of pronunciations) {
+                    const {pronunciation} = groupedPronunciation;
                     if (pronunciation.type !== 'pitch-accent') { continue; }
-                    const {terms, reading, position, nasalPositions, devoicePositions, tags, exclusiveTerms, exclusiveReadings} = pronunciation;
+                    const {terms, reading, exclusiveTerms, exclusiveReadings} = groupedPronunciation;
+                    const {position, nasalPositions, devoicePositions, tags} = pronunciation;
                     pitches.push({
                         expressions: terms,
                         reading,
@@ -225,9 +227,11 @@ export class AnkiNoteDataCreator {
         if (dictionaryEntry.type === 'term') {
             for (const {dictionary, pronunciations} of DictionaryDataUtil.getGroupedPronunciations(dictionaryEntry)) {
                 const phoneticTranscriptions = [];
-                for (const pronunciation of pronunciations) {
+                for (const groupedPronunciation of pronunciations) {
+                    const {pronunciation} = groupedPronunciation;
                     if (pronunciation.type !== 'phonetic-transcription') { continue; }
-                    const {terms, reading, ipa, tags, exclusiveTerms, exclusiveReadings} = pronunciation;
+                    const {ipa, tags} = pronunciation;
+                    const {terms, reading, exclusiveTerms, exclusiveReadings} = groupedPronunciation;
                     phoneticTranscriptions.push({
                         expressions: terms,
                         reading,
diff --git a/ext/js/dictionary/dictionary-data-util.js b/ext/js/dictionary/dictionary-data-util.js
index f889b8d0..408b1975 100644
--- a/ext/js/dictionary/dictionary-data-util.js
+++ b/ext/js/dictionary/dictionary-data-util.js
@@ -157,7 +157,7 @@ export class DictionaryDataUtil {
                 let groupedPronunciation = this._findExistingGroupedPronunciation(reading, pronunciation, dictionaryGroupedPronunciationList);
                 if (groupedPronunciation === null) {
                     groupedPronunciation = {
-                        ...pronunciation,
+                        pronunciation,
                         terms: new Set(),
                         reading
                     };
@@ -174,41 +174,19 @@ export class DictionaryDataUtil {
             /** @type {import('dictionary-data-util').GroupedPronunciation[]} */
             const pronunciations2 = [];
             for (const groupedPronunciation of dictionaryGroupedPronunciationList) {
-                const {terms, reading} = groupedPronunciation;
+                const {pronunciation, terms, reading} = groupedPronunciation;
                 const exclusiveTerms = !this._areSetsEqual(terms, allTerms) ? this._getSetIntersection(terms, allTerms) : [];
                 const exclusiveReadings = [];
                 if (multipleReadings) {
                     exclusiveReadings.push(reading);
                 }
-                switch (groupedPronunciation.type) {
-                    case 'pitch-accent': {
-                        const {position, nasalPositions, devoicePositions, tags} = groupedPronunciation;
-                        pronunciations2.push({
-                            type: 'pitch-accent',
-                            terms: [...terms],
-                            reading,
-                            position,
-                            nasalPositions,
-                            devoicePositions,
-                            tags,
-                            exclusiveTerms,
-                            exclusiveReadings
-                        });
-                        break;
-                    } case 'phonetic-transcription': {
-                        const {ipa, tags} = groupedPronunciation;
-                        pronunciations2.push({
-                            type: 'phonetic-transcription',
-                            terms: [...terms],
-                            reading,
-                            ipa,
-                            tags,
-                            exclusiveTerms,
-                            exclusiveReadings
-                        });
-                        break;
-                    }
-                }
+                pronunciations2.push({
+                    pronunciation,
+                    terms: [...terms],
+                    reading,
+                    exclusiveTerms,
+                    exclusiveReadings
+                });
             }
             results2.push({dictionary, pronunciations: pronunciations2});
         }
@@ -339,18 +317,19 @@ export class DictionaryDataUtil {
      * @returns {boolean}
      */
     static _arePronunciationsEquivalent(groupedPronunciation, pronunciation) {
-        if (!this._areTagListsEqual(groupedPronunciation.tags, pronunciation.tags)) {
+        const pronunciation0 = groupedPronunciation.pronunciation;
+        if (!this._areTagListsEqual(pronunciation0.tags, pronunciation.tags)) {
             return false;
         }
 
-        if (pronunciation.type === 'pitch-accent' && groupedPronunciation.type === 'pitch-accent') {
-            if (groupedPronunciation.position !== pronunciation.position ||
-                !this._areArraysEqual(groupedPronunciation.nasalPositions, pronunciation.nasalPositions) ||
-                !this._areArraysEqual(groupedPronunciation.devoicePositions, pronunciation.devoicePositions)) {
+        if (pronunciation.type === 'pitch-accent' && pronunciation0.type === 'pitch-accent') {
+            if (pronunciation0.position !== pronunciation.position ||
+                !this._areArraysEqual(pronunciation0.nasalPositions, pronunciation.nasalPositions) ||
+                !this._areArraysEqual(pronunciation0.devoicePositions, pronunciation.devoicePositions)) {
                 return false;
             }
-        } else if (pronunciation.type === 'phonetic-transcription' && groupedPronunciation.type === 'phonetic-transcription') {
-            if (groupedPronunciation.ipa !== pronunciation.ipa) {
+        } else if (pronunciation.type === 'phonetic-transcription' && pronunciation0.type === 'phonetic-transcription') {
+            if (pronunciation0.ipa !== pronunciation.ipa) {
                 return false;
             }
         } else {
diff --git a/ext/js/display/display-generator.js b/ext/js/display/display-generator.js
index 393092a5..d3f77883 100644
--- a/ext/js/display/display-generator.js
+++ b/ext/js/display/display-generator.js
@@ -621,7 +621,7 @@ export class DisplayGenerator {
         n1.appendChild(tag);
 
         let hasTags = false;
-        for (const {tags} of pronunciations) {
+        for (const {pronunciation: {tags}} of pronunciations) {
             if (tags.length > 0) {
                 hasTags = true;
                 break;
@@ -640,21 +640,24 @@ export class DisplayGenerator {
      * @returns {HTMLElement}
      */
     _createPronunciation(details) {
-        switch (details.type) {
+        const {pronunciation} = details;
+        switch (pronunciation.type) {
             case 'pitch-accent':
-                return this._createPronunciationPitchAccent(details);
+                return this._createPronunciationPitchAccent(pronunciation, details);
             case 'phonetic-transcription':
-                return this._createPronunciationPhoneticTranscription(details);
+                return this._createPronunciationPhoneticTranscription(pronunciation, details);
         }
     }
 
 
     /**
-     * @param {import('dictionary-data-util').GroupedPhoneticTranscription} details
+     * @param {import('dictionary').PhoneticTranscription} phoneticTranscription
+     * @param {import('dictionary-data-util').GroupedPronunciation} groupedPronunciation
      * @returns {HTMLElement}
      */
-    _createPronunciationPhoneticTranscription(details) {
-        const {ipa, tags, exclusiveTerms, exclusiveReadings} = details;
+    _createPronunciationPhoneticTranscription(phoneticTranscription, groupedPronunciation) {
+        const {ipa, tags} = phoneticTranscription;
+        const {exclusiveTerms, exclusiveReadings} = groupedPronunciation;
 
         const node = this._instantiate('pronunciation');
 
@@ -674,13 +677,15 @@ export class DisplayGenerator {
     }
 
     /**
-     * @param {import('dictionary-data-util').GroupedPitchAccent} details
+     * @param {import('dictionary').PitchAccent} pitchAccent
+     * @param {import('dictionary-data-util').GroupedPronunciation} groupedPronunciation
      * @returns {HTMLElement}
      */
-    _createPronunciationPitchAccent(details) {
+    _createPronunciationPitchAccent(pitchAccent, groupedPronunciation) {
         const jp = this._japaneseUtil;
 
-        const {reading, position, nasalPositions, devoicePositions, tags, exclusiveTerms, exclusiveReadings} = details;
+        const {position, nasalPositions, devoicePositions, tags} = pitchAccent;
+        const {reading, exclusiveTerms, exclusiveReadings} = groupedPronunciation;
         const morae = jp.getKanaMorae(reading);
 
         const node = this._instantiate('pronunciation');
diff --git a/types/ext/dictionary-data-util.d.ts b/types/ext/dictionary-data-util.d.ts
index f952da62..4ab06f11 100644
--- a/types/ext/dictionary-data-util.d.ts
+++ b/types/ext/dictionary-data-util.d.ts
@@ -63,51 +63,20 @@ export type KanjiFrequency = {
 
 export type TermFrequencyType = 'popular' | 'rare' | 'normal';
 
-export type GroupedPitchAccentInternal = {
-    type: 'pitch-accent';
+export type GroupedPronunciationInternal = {
+    pronunciation: Dictionary.Pronunciation;
     terms: Set<string>;
     reading: string;
-    position: number;
-    nasalPositions: number[];
-    devoicePositions: number[];
-    tags: Dictionary.Tag[];
 };
 
-export type GroupedPhoneticTranscriptionInternal = {
-    type: 'phonetic-transcription';
-    terms: Set<string>;
-    reading: string;
-    tags: Dictionary.Tag[];
-    ipa: string;
-};
-
-export type GroupedPronunciationInternal = GroupedPitchAccentInternal | GroupedPhoneticTranscriptionInternal;
-
-export type GroupedPitchAccent = {
-    type: 'pitch-accent';
+export type GroupedPronunciation = {
+    pronunciation: Dictionary.Pronunciation;
     terms: string[];
     reading: string;
-    position: number;
-    nasalPositions: number[];
-    devoicePositions: number[];
-    tags: Dictionary.Tag[];
     exclusiveTerms: string[];
     exclusiveReadings: string[];
 };
 
-export type GroupedPhoneticTranscription = {
-    type: 'phonetic-transcription';
-    terms: string[];
-    reading: string;
-    tags: Dictionary.Tag[];
-    ipa: string;
-    exclusiveTerms: string[];
-    exclusiveReadings: string[];
-};
-
-
-export type GroupedPronunciation = GroupedPitchAccent | GroupedPhoneticTranscription;
-
 export type DictionaryGroupedPronunciations = {
     dictionary: string;
     pronunciations: GroupedPronunciation[];
```